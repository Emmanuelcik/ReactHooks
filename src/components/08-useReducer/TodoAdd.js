import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({handleAddTodo}) => {

    const [ { description }, handleInputChange, reset ] = useForm({
        description: "",
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if(description.trim().length <= 1 ){
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false,
        }

        handleAddTodo(newTodo);
        reset();
    }
    console.log(description);
    return (
        <>
            <h4>Agregar ToDo</h4>
            <hr/>

            <form onSubmit={handleSubmit}>

                <input
                    className='form-control'
                    type="text"
                    name="description"
                    placeholder="Aprender..."
                    autoComplete='off'
                    onChange={handleInputChange}
                    value={description}
                />

                <button
                    type="submit"
                    className="btn btn-outline-primary d-block mt-2 col-12"
                > 
                    Agregar 
                </button>
            </form>
        </>
    )
}
