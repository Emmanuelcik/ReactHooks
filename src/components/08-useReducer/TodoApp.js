import React, {useEffect, useReducer} from 'react'
import { todoReducer } from './todoReducer';
import "./styles.css";
import { useForm } from '../../hooks/useForm';


const init = () => {

    let dataInLocal = localStorage.getItem("todos");
    return dataInLocal ? JSON.parse(dataInLocal) : [];

    // return [{
    //     id: new Date().getTime(),
    //     desc: "Learning react",
    //     done: false,
    // }];
}

export const TodoApp = () => {

    // const [state, dispatch] = useReducer(reducer, initialState, init)
    const [todos, dispatch] = useReducer(todoReducer, [], init);

    const [ {description}, handleInputChange, reset ] = useForm({
        description: "",
    });

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos));
    }, [todos]);

    const handleDelete = (todoID) => {
        const action = {
            type: "delete",
            payload: todoID,
        }

        dispatch(action);
    }
    
    console.log(description);

    const handleSubmit = (e) =>  {
        e.preventDefault();

        if(description.trim().length <= 1 ){
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false,
        }

        const action = {
            type: "add",
            payload: newTodo
        }

        dispatch(action);
        reset();
    }

    return (
        <div>
            <h1>Todo App ({todos.length})</h1>
            <hr/>
            <div className="row">

                <div className="col col-md-6">
                    <ul className="list-group list-group-flush">
                        {
                            todos.map( (todo, i) => (
                                <li
                                    key={todo.id}
                                    className="list-group-item"
                                >
                                    <p className='text-center'>{i + 1}.- { todo.desc} </p>
                                    <button
                                        onClick={() => { handleDelete( todo.id)}}
                                        className="btn btn-danger"
                                    >
                                        Borrar
                                    </button>
                                </li>
                            )) 
                        }
                    </ul>
                </div>
                <div className='col col-md-6'>
                    <h4>Agregar ToDo</h4>
                    <hr/>

                    <form onSubmit={handleSubmit}>

                        <input
                            className='form-control'
                            type="text"
                            name="description"
                            placeholder="Aprender..."
                            autoComplete='off'
                            onChange={handleInputChange}
                            value={description}
                        />

                        <button
                            type="submit"
                            className="btn btn-outline-primary d-block mt-2"
                        > 
                            Agregar 
                        </button>
                    </form>

                </div>

            </div>

        </div>
    )
}
