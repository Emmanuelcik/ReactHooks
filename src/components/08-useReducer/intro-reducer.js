

const initialState = [{
    id: 1,
    todo: "Comprar leche",
    done: false,    
}];

const toDoReducer = (state = initialState, action) => {
    
    if(action?.type === "agregar"){
        return [...state, action.payLoad];
    }

    return state;
}

let todos = toDoReducer();

const newTodo = {
    id: 2,
    todo: "Comprar leche",
    done: false,
}

const action = {
    type: "agregar",
    payLoad: newTodo,
    // newTodo
}

todos = toDoReducer(todos, action);
console.log(todos);
