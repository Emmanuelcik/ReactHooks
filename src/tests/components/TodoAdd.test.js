import { shallow } from "enzyme";
import { TodoAdd } from "../../components/08-useReducer/TodoAdd";


describe("Testing TodoAdd", () => {

    const handleAddTodo = jest.fn();
    const wrapper = shallow( 
        <TodoAdd 
            handleAddTodo={handleAddTodo}
        /> 
    )

    test("Should match the snapshot", () => {

        expect(wrapper).toMatchSnapshot();
    });
    test("Should not call handleAddTodo", () => {

        const formSubmit = wrapper.find("form").prop("onSubmit");

        formSubmit({ preventDefault(){} });

        expect(handleAddTodo).toHaveBeenCalledTimes(0);
    });
    test("Should call handleAddTodo", () => {

        const value = "Aprender react";

        wrapper.find("input").simulate("change", {
            target : {
                value,
                name: "description"
            }
        });

        const formSubmit = wrapper.find("form").prop("onSubmit");

        formSubmit({ preventDefault(){} });

        expect(handleAddTodo).toHaveBeenCalledTimes(1);
        expect(handleAddTodo).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false
        })
    });
});