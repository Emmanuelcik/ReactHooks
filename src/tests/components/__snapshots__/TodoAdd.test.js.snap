// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing TodoAdd Should match the snapshot 1`] = `
<Fragment>
  <h4>
    Agregar ToDo
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprender..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary d-block mt-2 col-12"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
